Userland/unzip.cpp

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2020, Andrés Vieira <anvieiravazquez@gmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <LibCore/ArgsParser.h>
+#include <LibCore/File.h>
+#include <string.h>
+#include <sys/stat.h>
+
+static const u8 central_directory_file_header_sig[] = "\x50\x4b\x01\x02";
+
+bool find_next_central_directory(off_t file_size, Core::File* file, off_t current_index, off_t* return_index)
+{
+    u8 buffer[4];
+    off_t start_index = current_index == 0 ? current_index : current_index + 1;
+    for (off_t index = start_index; index < file_size - 4; index++) {
+        file->seek(index);
+        file->read(buffer, 4);
+        if (buffer[0] == central_directory_file_header_sig[0] && !memcmp(buffer, central_directory_file_header_sig, 4)) {
+            *return_index = index;
+            return true;
+        }
+    }
+    return false;
+}
+
+bool unpack_file_for_central_directory_index(off_t central_directory_index, Core::File* file)
+{
+    u8 buffer[4];
+
+    off_t local_file_header_index;
+    file->seek(central_directory_index + 42);
+    file->read(buffer, 4);
```

Comment:
Missing error handling (same comment repeated many times.)
I bet we could make some kind of helper that does e.g "seek to P and read N bytes" :)